AdaptiveMaxPool3d tests need to generate non-equal values

Intermittent test failures occurring on AdaptiveMaxPool3d.

@gchanan and I have seen this in the past where Max pooling has exactly the same values in the same window, then it's differently resolved on CPU/GPU. For example [0 2.5 2.5 3] can give a max index of `1` or `2` on CUDA depending on the runtime.

So generating input values that are within epsilon of each other is important (especially at half precision).

@SsnL fix that one.

```
======================================================================
FAIL: test_AdaptiveMaxPool3d_tuple (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_nn.py", line 3477, in <lambda>
    setattr(TestNN, test_name, lambda self, test=test: test(self))
  File "<https://build.pytorch.org/job/pytorch-master-py2-linux/ws/builder/pytorch/test/common_nn.py",> line 631, in __call__
    self._do_test(test_case, module, input)
  File "test_nn.py", line 109, in _do_test
    test_case.check_jacobian(module, input, self.jacobian_input)
  File "<https://build.pytorch.org/job/pytorch-master-py2-linux/ws/builder/pytorch/test/common_nn.py",> line 519, in check_jacobian
    PRECISION
AssertionError: 0.4831197783494723 not less than or equal to 1e-05

======================================================================
FAIL: test_AdaptiveMaxPool3d_tuple_cuda (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_nn.py", line 3478, in <lambda>
    setattr(TestNN, cuda_test_name, lambda self, test=test: test.test_cuda(self))
  File "<https://build.pytorch.org/job/pytorch-master-py2-linux/ws/builder/pytorch/test/common_nn.py",> line 711, in test_cuda
    test_case.assertEqual(cpu_gradInput, gpu_gradInput, 2e-4)
  File "<https://build.pytorch.org/job/pytorch-master-py2-linux/ws/builder/pytorch/test/common.py",> line 215, in assertEqual
    assertTensorsEqual(x, y)
  File "<https://build.pytorch.org/job/pytorch-master-py2-linux/ws/builder/pytorch/test/common.py",> line 207, in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
AssertionError: 2.0 not less than or equal to 0.0002
```

--- a/test/test_nn.py
+++ b/test/test_nn.py
@@ -3466,6 +3466,12 @@ class TestNNInit(TestCase):
                         self.assertEqual(torch.mm(flattened_tensor, flattened_tensor.t()),
                                          torch.eye(rows) * gain ** 2, prec=1e-6)
 
+# Generates rand tensor with non-equal values. This ensures that duplicate
+# values won't be causing test failure for modules like MaxPooling.
+# size should be small, otherwise randperm fails / long overflows.
+def _rand_tensor_non_equal(*size):
+    total = reduce(mul, size, 1)
+    return torch.randperm(total).view(*size).double()
 
 def add_test(test):
     test_name = test.get_name()
@@ -4138,42 +4144,42 @@ new_module_tests = [
     dict(
         module_name='AdaptiveMaxPool1d',
         constructor_args=(3,),
-        input=torch.rand(1, 3, 5),
+        input=_rand_tensor_non_equal(1, 3, 5),
     ),
     dict(
         module_name='AdaptiveMaxPool2d',
         constructor_args=(3,),
-        input=torch.rand(1, 3, 5, 6),
+        input=_rand_tensor_non_equal(1, 3, 5, 6),
         desc='single',
     ),
     dict(
         module_name='AdaptiveMaxPool2d',
         constructor_args=((3, 4),),
-        input=torch.rand(1, 3, 5, 6),
+        input=_rand_tensor_non_equal(1, 3, 5, 6),
         desc='tuple',
     ),
     dict(
         module_name='AdaptiveMaxPool3d',
         constructor_args=(3,),
-        input=torch.rand(2, 3, 5, 6, 7),
+        input=_rand_tensor_non_equal(2, 3, 5, 6, 7),
         desc='single',
     ),
     dict(
         module_name='AdaptiveMaxPool3d',
         constructor_args=((3, 4, 5),),
-        input=torch.rand(2, 3, 5, 6, 7),
+        input=_rand_tensor_non_equal(2, 3, 5, 6, 7),
         desc='tuple',
     ),
     dict(
         module_name='AdaptiveMaxPool3d',
         constructor_args=(3,),
-        input=torch.rand(2, 3, 12, 9, 3),
+        input=_rand_tensor_non_equal(2, 3, 12, 9, 3),
         desc='single_nonatomic',
     ),
     dict(
         module_name='AdaptiveMaxPool3d',
         constructor_args=((3, 4, 5),),
-        input=torch.rand(2, 3, 6, 4, 10),
+        input=_rand_tensor_non_equal(2, 3, 6, 4, 10),
         desc='tuple_nonatomic',
     ),
     dict(

